packages/experiments/src/components/Toggle/Toggle.state.ts

```diff
@@ -1,51 +1,51 @@
-import * as React from 'react';
-import { IToggle, IToggleProps, IToggleViewProps } from './Toggle.types';
-import { BaseState } from '../../utilities/BaseState';
+import { useCallback, useImperativeHandle, useRef } from 'react';
+import { getControlledDerivedProps, useControlledState } from '../../Foundation';
+import { IToggleComponent, IToggleViewProps } from './Toggle.types';
 
-export type IToggleState = Pick<IToggleViewProps, 'checked' | 'onChange' | 'onClick' | 'text' | 'toggleButtonRef'>;
+export const useToggleState: IToggleComponent['state'] = props => {
+  const toggleButtonRef = useRef<HTMLButtonElement | null>(null);
 
-export class ToggleState extends BaseState<IToggleProps, IToggleViewProps, IToggleState> implements IToggle {
-  private _toggleButtonRef = React.createRef<HTMLButtonElement>();
+  const [checked, setChecked] = useControlledState(props, 'checked', { defaultPropName: 'defaultChecked', defaultPropValue: false });
 
-  constructor(props: ToggleState['props']) {
-    super(props, {
-      controlledProps: ['checked', 'text'],
-      transformViewProps: (newProps: IToggleViewProps) => {
-        newProps.text = newProps.checked ? props.onText : props.offText;
-        return newProps;
-      }
-    });
-
-    this.state = {
-      checked: !!props.defaultChecked,
-      text: !!props.defaultChecked ? props.onText : props.offText,
-      onChange: this._noop,
-      onClick: this._onClick,
-      toggleButtonRef: this._toggleButtonRef
-    };
-  }
-
-  public focus = () => {
-    if (this._toggleButtonRef.current) {
-      this._toggleButtonRef.current.focus();
+  useImperativeHandle(props.componentRef, () => ({
+    focus: () => {
+      toggleButtonRef.current && toggleButtonRef.current.focus();
     }
-  };
+  }));
 
-  private _onClick = (ev: React.MouseEvent<HTMLElement>) => {
-    const { disabled, onChange } = this.props;
-    const { checked } = this.state;
+  const { disabled, onChange } = props;
 
-    if (!disabled) {
-      // Only update the state if the user hasn't provided it.
-      this.setState({ checked: !checked });
+  const _onClick = useCallback(
```

Comment:
still wrapping my brain on this... it seems that the function will be re-created on every render (but the useCallback will return the original value, I guess? for immutability purposes.)

Reply:
I was wondering if we could use some kind of closure wrapper that encapsulates new function creation, passes in the dep values and only gets called when memoize retrieval fails. However, React states that there is no guarantee that previously memoized values will remain available, so we'd need some kind of hook to know when React is creating a new copy.